src/api/modelTypes.js

```diff
@@ -242,6 +242,10 @@ export type CrossRealmBot = {|
   // We assume this is `true` when present.
   // TODO(server-5.0): Replaced in FL 83 by is_system_bot
   +is_cross_realm_bot?: true,
+
+  // A cross-realm bot has no custom profile fields set.
+  // But there doesn't seem to be a need to enforce that in the type.
+  +profile_data?: $ElementType<User, 'profile_data'>,
```

Comment:
> I think there were two goals in that commit:
> 
> * Find out what the API is; don't just write what we see servers doing
> * Match our types more precisely to the API
> 
> These are good goals that can help us make code less complicated and error-prone.

Yep! I agree.



> This change is helpful because it lets us treat `User`s and `CrossRealmBot`s more uniformly, in a UI that doesn't "switch/case" on whether it's rendering a user or a bot.

Yeah. In general I think having the separate type `CrossRealmBot` is a legacy wart -- the distinction between human users and normal per-realm bots on the one hand (typed as `User`), and cross-realm bots on the other hand, is pretty much an implementation detail as far as we're concerned, and shouldn't have so much prominence in our code.

Before your recent ab4d74683f8a5df9a31902162f2a45ab7e1edb73 and 9325bd2a72b7d62335a4980d543c160156510954, there were a lot of small differences between the types, and so unifying them was a cleanup task I'd never quite finished. (In 2018 into 2019 there was a lot of confusion about this in our code, which starting with 8ef11eb7663d40a659bd5f06027c87dbfa347095 I cleaned up a lot of but not all.) But now they're very similar! So it might be a reasonably short push from here to (a) use `UserOrBot` everywhere, and then (b) make that be just a single object type that covers both, and call it `User`.



> And it's pretty harmless because we never construct `CrossRealmBot`s ourselves,

Right. These types purely describe data we get from the server -- they don't describe expectations the server has for any data it gets from us. So we can freely make the types less specific, and that's always safe. (At worst it could mean a test case supplies invalid data and we don't notice, making the test less helpful.)

Reply:
> * I think this should replace the `profile_data is never present` line, earlier in the list of properties.

Thanks, yeah, I'd missed that line.


> * I'm not sure it makes a difference, but the doc says bots _can't_ have custom profile fields; not that they _don't_.

Yeah, in this context I think those are different ways of saying the same thing.


> * Without more context, is it clear that "A cross-realm bot has no custom profile fields set" means that the `profile_data` property is absent? (As opposed to present with an empty object, or with `null`, etc.) …

No, I think it leaves it ambiguous between absent and empty-object. (The type doesn't admit `null`.)

The API docs leave the same question not quite clear when it comes to `realm_users` and `realm_non_active_users` too. They say `profile_data` is an object and not optional, so taken literally that should mean it's always present (and would just be an empty object when no fields set)… but then they say it's absent for bots, so one can't take it literally. Further, they tie its absence for bots to their not having custom profile fields, which seems to actively suggest that it might be absent for other users that have no fields too.


> … I think saying this in the comment could help the reader understand what "enforce that in the type" means. (I guess the doc does say it though; maybe that's enough.)

Yeah. Quite possibly I should just delete the comment.

Reply:
Cool, makes sense, thanks for those explanations.